### PR TITLE
fix: expand cargo key scope room invitations

### PR DESCRIPTION
**Pull request by roman-khomenko-dev**
Add cargo room dialogs and invitations up to 1 handshake from operator to backup

### CODE REVIEW

File: lib/chat/db/scope/utils.ex [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  def get_dialog_binkeys(dialog_keys),
+    do: dialog_keys |> Enum.map(fn {:dialogs, dialog_key} -> dialog_key end) |> MapSet.new()
+
+  def get_dialog_invitation_messages(dialog_binkeys, snap) do
+    snap
+    |> db_stream({:dialog_message, 0, 0, 0}, {:"dialog_message\0", 0, 0, 0})
+    |> Stream.filter(fn {{:dialog_message, key, _, _}, %Message{type: type}} ->
+      type == :room_invite and MapSet.member?(dialog_binkeys, key)
+    end)
+  end
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/utils.ex:90]: Read
-  https://hexdocs.pm/cubdb/CubDB.html#select/2 
-  https://hexdocs.pm/elixir/1.15/operators.html#term-ordering

We may select messages of dialogs we interested in and filter out by type. With something like
```elixir
dialog_binkeys
|> Stream.map(fn dialog_key -> {{:dialog_message, dialog_key, 0, 0}, {:dialog_message, dialog_key, nil, nil}} end)
|> Stream.map(fn range ->
  snap
  |> db_stream(range)
  |> Stream.filter(& match?({_, %Mesaage{type:  :room_invite}, &1))
  |> just_keys() 
end)
|> List.flatten()
```

Check where this approach can be applied

File: lib/chat/db/scope/key_scope.ex [hunk 2]
```diff
@@ -132,47 +132,60 @@ defmodule Chat.Db.Scope.KeyScope do
...
+      [checkpoints_invitation_messages, operator_invitation_messages, nested_invitation_messages]
+      |> Stream.concat()
+      |> Stream.map(&Utils.just_keys/1)
+      |> union_set_dialog_keys([
+        checkpoints_dialog_keys,
+        operator_dialog_keys,
+        nested_dialog_keys
+      ])
+      |> Utils.union_set(acc_set),
+      MapSet.union(pub_keys, nested_user_key)
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/key_scope.ex:180]: It is hard to understand what happens here

may be start with comments

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: test/chat/db/scope/cargo_key_scope_test.exs [hunk 3]
```diff
@@ -1,41 +1,129 @@
...
+    room_identity
+  end
+
+  defp generate_dialogs_and_cargo_room_invite(checkpoints, user, room_identity) do
+    dialogs =
+      for checkpoint <- checkpoints,
+          into: [],
+          do: Dialogs.find_or_open(user, checkpoint |> Chat.Card.from_identity())
+
+    Enum.each(dialogs, fn dialog -> create_and_add_room_invite(room_identity, dialog, user) end)
```
Comment (sergey-lukianov) [inline test/chat/db/scope/cargo_key_scope_test.exs:111]: `for` and `Enum.each`  doing the same

Why to use both ?

File: test/chat/db/scope/cargo_key_scope_test.exs [hunk 4]
```diff
@@ -1,41 +1,129 @@
...
+
+  defp store_cargo_settings_checkpoints(cards) do
+    :ok =
+      AdminRoom.store_cargo_settings(%{
+        AdminRoom.get_cargo_settings()
+        | checkpoints: cards
+      })
+  end
+
+  defp register_and_put_user(identity) do
```
Comment (sergey-lukianov) [inline test/chat/db/scope/cargo_key_scope_test.exs:96]: bad name

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: lib/chat/db/scope/key_scope.ex [hunk 5]
```diff
@@ -3,8 +3,8 @@ defmodule Chat.Db.Scope.KeyScope do
   Builds db keys accessible by keys_list
   """
 
+  alias Chat.Db.Scope.Utils
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex]: right, done

File: lib/chat/db/scope/key_scope.ex (hunk 2 above)
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex]: each step commented

File: test/chat/db/scope/cargo_key_scope_test.exs (hunk 4 above)
Comment (roman-khomenko-dev) [reply test/chat/db/scope/cargo_key_scope_test.exs]: renamed to create_user_from_identity

File: test/chat/db/scope/cargo_key_scope_test.exs (hunk 3 above)
Comment (roman-khomenko-dev) [reply test/chat/db/scope/cargo_key_scope_test.exs]: right, function from each moved to for

File: test/chat/db/scope/cargo_key_scope_test.exs [hunk 6]
```diff
@@ -1,41 +1,129 @@
...
+    generate_dialogs_and_cargo_room_invite([admin, john], operator, room_identity)
+    user_c = User.login("UserC") |> create_user_from_identity()
+    generate_dialogs_and_cargo_room_invite([user_c], operator, room_identity)
+
+    user_d = User.login("UserD") |> create_user_from_identity()
+    generate_dialogs_and_cargo_room_invite([user_d], user_c, room_identity)
+
+    assert_keys_for_cargo_keys(room_key, [checkpoint1_key, checkpoint2_key], 4)
+  end
+
```
Comment (sergey-lukianov) [inline test/chat/db/scope/cargo_key_scope_test.exs:43]: add a test with 
- checkpoints c1,c2,c3
- operarators o1, o2
- users u1,u2,u3, u4

room was created by o1 with checkpoints list c1 and c2
later o2 added manual invite for c3 into room
o1 invited u1 
u1 invited u2, u4
u2 invited u3
u3 invited u4 and u1

u3's invite should not get into scope

File: lib/chat/db/scope/key_scope.ex [hunk 7]
```diff
@@ -143,36 +143,57 @@ defmodule Chat.Db.Scope.KeyScope do
...
+      # merge dialog keys into acc_set
+      |> union_set_dialog_keys([
+        checkpoints_dialog_keys,
+        operator_dialog_keys,
+        nested_dialog_keys
+      ])
+      |> union_set(acc_set),
+      # combine keys of checkpoints and nested users to get complete list of invite content
+      MapSet.union(pub_keys, nested_user_key)
+    }
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex:189]: nested user is level one and we get dialogs and invitations from them to level2 in code, ill recheck it

File: lib/chat/db/scope/key_scope.ex [hunk 8]
```diff
@@ -143,36 +143,57 @@ defmodule Chat.Db.Scope.KeyScope do
...
-      |> Stream.filter(fn {_full_dilaog_key, %Dialog{a_key: a_key, b_key: b_key}} ->
-        MapSet.member?(pub_keys, a_key) or MapSet.member?(pub_keys, b_key)
-      end)
+    # get all dialogs
+    dialogs = get_dialogs(snap)
+    # select dialog keys and invitation messages to checkpoints
+    {checkpoints_dialog_keys, checkpoints_invitation_messages} =
+      extract_dialogs_with_invitations(:checkpoints, snap, dialogs, [pub_keys, nil])
+
+    # get operator key as first sender of invitations to checkpoints
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex:152]: yeah, i just write comment it in a single

[the remark this answers is not in the document]
Comment (sergey-lukianov) [reply lib/chat/db/scope/key_scope.ex:152]: and variable name in singular

File: lib/chat/db/scope/utils.ex (hunk 1 above)
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/utils.ex]: this function and get_invitations_sender_keys refactored according to select from cubedb

File: CHANGELOG.md [hunk 9]
```diff
@@ -22,6 +22,8 @@ straightforward as possible.
 
 ### Fixed
 
+- user access to cargo room
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply CHANGELOG.md]: changed

File: lib/chat/db/scope/key_scope.ex (hunk 8 above)
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex]: naming changed to many

File: lib/chat/db/scope/key_scope.ex (hunk 7 above)
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex:189]: all going right :)

[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex:189]: change functions to make one pipe flow soon by comments

File: lib/chat/db/scope/utils.ex (hunk 1 above)
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/utils.ex]: iv see more approach, change it soon

File: test/chat/db/scope/cargo_key_scope_test.exs (hunk 6 above)
Comment (roman-khomenko-dev) [reply test/chat/db/scope/cargo_key_scope_test.exs:82]: test rewrited with dynamic setup and described case

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: lib/chat/db/scope/utils.ex [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+      :recipient -> if(is_a_to_b, do: dialog.b_key, else: dialog.a_key)
+    end
+  end
+
+  defp fetch_dialog_by_key({snap, dialog_key}),
+    do:
+      snap
+      |> db_stream({:dialogs, dialog_key}, {:"dialogs\0", dialog_key})
+      |> Enum.to_list()
+      |> List.first()
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/utils.ex:101]:
```suggestion
     |> CubDB.Snapshot.get({:dialogs, dialog_key})
```

https://hexdocs.pm/cubdb/CubDB.Snapshot.html#get/3

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: lib/chat/db/scope/key_scope.ex (hunk 7 above)
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex:166]: flow piped

File: test/chat/db/scope/cargo_key_scope_test.exs [hunk 11]
```diff
@@ -1,41 +1,174 @@
...
+    generate_dialogs_and_cargo_room_invite([Enum.at(users, 2)], Enum.at(users, 1), room_identity)
+
+    generate_dialogs_and_cargo_room_invite(
+      [Enum.at(users, 3), Enum.at(users, 0)],
+      Enum.at(users, 2),
+      room_identity
+    )
+
+    assert_keys_for_cargo_keys(room_key, checkpoint_keys, 7)
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply test/chat/db/scope/cargo_key_scope_test.exs]: but have some failed now with it

File: lib/chat/db/scope/key_scope.ex [hunk 12]
```diff
@@ -143,36 +143,34 @@ defmodule Chat.Db.Scope.KeyScope do
...
+
+    {
+      context_info.messages
+      |> get_messages_keys()
+      |> merge_dialogs_keys(
+        acc_set,
+        context_info.dialogs_keys
+      ),
+      full_hadshake_keys(pub_keys, context["nested_users_keys"])
+    }
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/key_scope.ex:166]: A glimpse  of readability :)

It was returnig a MapSet before, now a tuple. Is it ok ?

Comment (sergey-lukianov) [reply lib/chat/db/scope/key_scope.ex:166]: And this is bruteforce variation.

Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex:166]: I change it to pass full needed keys into add_invitation_content func which is next step of backup creation

File: test/chat/db/scope/cargo_key_scope_test.exs [hunk 13]
```diff
@@ -1,41 +1,206 @@
...
+    generate_dialogs_and_cargo_room_invite([user_c], operator, room_identity)
+
+    user_d = List.last(users)
+    generate_dialogs_and_cargo_room_invite([user_d], user_c, room_identity)
+
+    assert_keys_for_cargo_keys(room_key, checkpoint_keys, 4)
+  end
+
+  test "room invitations found up to 1 handshake" do
+    {[checkpoints, checkpoint_keys], operators, users} = setup_test_data(3, 2, 4)
```
[the remark this answers is not in the document]
Comment (roman-khomenko-dev) [reply test/chat/db/scope/cargo_key_scope_test.exs:49]: right, thx

File: lib/chat/db/scope/key_scope.ex (hunk 12 above)
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex]: recursive variation done

File: lib/chat/db/scope/key_scope.ex [hunk 14]
```diff
@@ -28,8 +28,9 @@ defmodule Chat.Db.Scope.KeyScope do
       |> add_rooms(snap, room_key)
       |> add_content(snap, room_key)
       |> add_invitation_dialogs(snap, invited_keys)
-      |> add_invitation_content(snap, invited_keys)
     end)
+
+    # |> IO.inspect(label: "cargo keys")
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/key_scope.ex:33]: remove

File: lib/chat/db/scope/key_scope.ex [hunk 15]
```diff
@@ -3,8 +3,8 @@ defmodule Chat.Db.Scope.KeyScope do
   Builds db keys accessible by keys_list
   """
 
+  import Chat.Db.Scope.{Utils, InvitationsHandshaker}
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/key_scope.ex:6]: no {} in aliases please

File: lib/chat/db/scope/key_scope.ex [hunk 16]
```diff
@@ -143,92 +144,13 @@ defmodule Chat.Db.Scope.KeyScope do
...
-    dialog_keys =
-      snap
-      |> db_stream({:dialogs, 0}, {:"dialogs\0", 0})
-      |> Stream.filter(fn {_full_dilaog_key, %Dialog{a_key: a_key, b_key: b_key}} ->
-        MapSet.member?(pub_keys, a_key) or MapSet.member?(pub_keys, b_key)
-      end)
-      |> Stream.map(&just_keys/1)
-      |> Enum.to_list()
-      |> MapSet.new()
+    context =
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/key_scope.ex:147]: this can be one pipe in this funciton

File: lib/chat/db/scope/invitations_handshaker.ex [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+
+  def build_initial_context(snap) do
+    %{}
+    |> put_room_invitations_content(snap, nil)
+    |> put_invitations_keymap()
+    |> put_invitations_messages(snap)
+    |> clear_context_preparation()
+  end
+
+  def put_room_invitations_content(context, snap, keys) do
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/invitations_handshaker.ex:18]: all the function in this module are `def`s  
Can we make `defp` for internal ones

File: lib/chat/db/scope/invitations_handshaker.ex [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+        |> fetch_cycle_invites(context)
+        |> put_invitations_queue(context)
+        |> remove_queued_invitations_messages()
+      else
+        context
+      end
+
+    move = move + 1
+    if move == 3, do: context, else: execute_handshake_cycle(context, move)
+  end
```
Comment (sergey-lukianov) [inline lib/chat/db/scope/invitations_handshaker.ex:83]: What is the logic behind `move`?

File: lib/chat/db/scope/key_scope.ex (hunk 14 above)
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/key_scope.ex]: oops, done

File: lib/chat/db/scope/invitations_handshaker.ex (hunk 17 above)
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/invitations_handshaker.ex]: right, fixed

File: lib/chat/db/scope/invitations_handshaker.ex (hunk 18 above)
Comment (roman-khomenko-dev) [reply lib/chat/db/scope/invitations_handshaker.ex]: Move is describe a current cycling number of getting handshake invites. Based on it i get messages direction and catch when should cycle be stopped. It starts from 0 and get invitations messages to checkpoints, then use its senders for each next iteration

Comment (roman-khomenko-dev) [reply lib/chat/db/scope/invitations_handshaker.ex]: Ok i find a way to not use this magic variable

Comment (roman-khomenko-dev) [reply lib/chat/db/scope/invitations_handshaker.ex]: Refactored

File: test/chat/db/scope/cargo_key_scope_test.exs [hunk 19]
```diff
@@ -1,41 +1,178 @@
...
+
+    invitations_board =
+      %{}
+      |> generate_cargo_room(operator_1)
+      |> put_dialogs_with_invites([Enum.slice(checkpoints, 0..1), operator_1], :index_1)
+      |> put_dialogs_with_invites([[Enum.at(checkpoints, 2)], operator_2], :index_2)
+      |> put_dialogs_with_invites([[user_1], operator_1], :index_3)
+      |> put_dialogs_with_invites([[user_2, user_4], user_1], :index_4)
+      |> put_dialogs_with_invites([[user_3], user_2], :index_5)
+      |> put_dialogs_with_invites([[user_4, user_1], user_3], :index_6)
```
Comment (sergey-lukianov) [inline test/chat/db/scope/cargo_key_scope_test.exs:65]: `|> send_invite(from: [user_4, user_1], to: user_3, mark_as: :index_6)`

This looks more readable

File: test/chat/db/scope/cargo_key_scope_test.exs [hunk 20]
```diff
@@ -1,41 +1,178 @@
...
+      [user_1, user_2, user_3, user_4]
+    } = setup_test_data(3, 2, 4)
+
+    assert Enum.count(checkpoints) == 3
+
+    invitations_board =
+      %{}
+      |> generate_cargo_room(operator_1)
+      |> put_dialogs_with_invites([Enum.slice(checkpoints, 0..1), operator_1], :index_1)
+      |> put_dialogs_with_invites([[Enum.at(checkpoints, 2)], operator_2], :index_2)
```
Comment (sergey-lukianov) [inline test/chat/db/scope/cargo_key_scope_test.exs:61]:
```suggestion
      |> put_dialogs_with_invites([checkpoint_1, checkpoint_2], operator_1], :index_1)
      |> put_dialogs_with_invites([[checkpoint_3], operator_2], :index_2)
```

Comment (roman-khomenko-dev) [reply test/chat/db/scope/cargo_key_scope_test.exs]: yep

File: CHANGELOG.md [hunk 21]
```diff
@@ -26,6 +26,8 @@ straightforward as possible.
 - mobile upload on Android
 - device log. new format support [#565]
 
+- copy invited user invitation with cargo room [#574]
```
Comment (sergey-lukianov) [inline CHANGELOG.md:29]: into unreleased one please

**Review by sergey-lukianov — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 21 above)
Comment (roman-khomenko-dev) [reply CHANGELOG.md]: updated

**Review by sergey-lukianov — APPROVED** [review]

**Pull request merged**